File: problems/ants/haskell/ants.hs
-- kattis-accepted
module Main where

import Data.List (sort)

main :: IO ()
main = interact solve

solve :: String -> String
solve = unlines . solveNum . tail . map read . words

solveNum :: [Int] -> [String]
solveNum [] = []
solveNum (len:count:xs) = oneAnswer : solveNum (drop count xs)
  where
    oneAnswer = solveOnce len $ take count xs

solveOnce :: Int -> [Int] -> String
solveOnce len ants = (unwords . map show)
  [ findShortest len ants
  , findLongest  len ants
  ]

findShortest :: Int -> [Int] -> Int
findShortest len = maximum . map (\a -> min a (len - a))

findLongest :: Int -> [Int] -> Int
findLongest len = maximum . map (\a -> max a (len - a))

File: problems/ants/python3/ants.py
# kattis-accepted
def main():
    num_cases = int(input())
    for case in range(num_cases):
        length, num_ants = map(int, input().split())
        ants = []
        while len(ants) < num_ants:
            ants.extend(map(int, input().split()))
        ants.sort()
        solve(ants, length)


def solve(ants, length):
    shortest = 0
    for ant in ants:
        dist = min(ant, length - ant)
        shortest = max(shortest, dist)
    longest = 0
    for ant in ants:
        dist = max(ant, length - ant)
        longest = max(longest, dist)
    print(shortest, longest)


if __name__ == "__main__":
    main()
